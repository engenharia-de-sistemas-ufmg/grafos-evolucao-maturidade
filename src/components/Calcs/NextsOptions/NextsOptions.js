/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { nextsType, graphType } from '../../../utils/types';
import generateRandom from '../../../utils/random';
import './NextsOptions.scss';

const NextsOptions = ({
  graph,
  nexts,
  showNexts,
  hideNexts,
  beginCalcNexts,
  changeNodesStatus,
}) => {
  let content = null;
  const [_minWeight, setMinWeight] = useState(195);
  const [_maxWeight, setMaxWeight] = useState(375);
  const [_maxDistance, setMaxistance] = useState(3);
  const [_forceMinimumLevel, setForceMinimumLevel] = useState(true);
  const [_showNextsOptions, setShowNextsOptions] = useState([]);
  const [_mustIncludeNodes, setMustIncludeNodes] = useState('');
  const [_mustNotIncludeSlots, setMustNotIncludeSlots] = useState('');

  if (nexts.calculated) {
    content = (
      <div>
        {nexts.options.length === 0 ? <div>Nenhuma combinação encontrada </div> : null}
        {nexts.options.map((o, i) => (
          <div className="card" key={generateRandom()}>
            <header
              className="card-header"
              onClick={() => setShowNextsOptions(
                _showNextsOptions.map((v, i1) => (i1 === i ? !v : v)),
              )}
            >
              <p className="card-header-title">Opção {i + 1}</p>
              <p className="card-header-subtitle">Distância: <b>{o.distance}</b>, Peso: <b>{o.totalWeight}</b>, Slots: <b>{o.slots.join(',')}</b></p>
            </header>
            {_showNextsOptions[i]
              && (
                <div className="card-content">
                  <ul className="inner-ul">
                    {o.combination.map((c) => (
                      <li key={generateRandom()}>
                        {c.label} - {c.title} <b>{c.level}</b> - <b>{c.weight}</b> - <b>{c.slots.join(',')}</b>
                      </li>
                    ))}
                  </ul>
                  <div className="is-divider" />
                  <div className="select-button-container">
                    <button
                      type="button"
                      className="button is-primary is-small"
                      onClick={() => { changeNodesStatus(o.combination.map((c) => c.id)); }}
                    >
                      Selecionar
                    </button>
                  </div>
                </div>
              )}
          </div>
        ))}
        <button
          className="button is-text is-small"
          type="button"
          onClick={() => { showNexts(); }}
        >
          Voltar
        </button>
      </div>
    );
  } else if (nexts.calculating) {
    content = (<div>Calculando ...</div>);
  } else {
    content = (
      <div className="content">
        <div className="columns is-paddingless is-marginless">
          <div className="field column is-one-quarter">
            <label className="label is-small" htmlFor="minWeight">Peso mínimo
              <div className="control">
                <input
                  className="input is-small"
                  type="text"
                  id="minWeight"
                  placeholder="195"
                  value={_minWeight}
                  onChange={(e) => { setMinWeight(e.target.value); }}
                />
              </div>
            </label>
          </div>
          <div className="field column is-one-quarter">
            <label className="label is-small" htmlFor="maxWeight">Peso máximo
              <div className="control">
                <input
                  className="input is-small"
                  type="text"
                  id="maxWeight"
                  placeholder="375"
                  value={_maxWeight}
                  onChange={(e) => { setMaxWeight(e.target.value); }}
                />
              </div>
            </label>
          </div>
          <div className="field column is-one-quarter">
            <label className="label is-small" htmlFor="maxDistance">Distância Máxima
              <div className="control">
                <input
                  className="input is-small"
                  type="text"
                  id="maxDistance"
                  placeholder="3"
                  value={_maxDistance}
                  onChange={(e) => { setMaxistance(e.target.value); }}
                />
              </div>
            </label>
          </div>
          <div className="field column is-one-quarter">
            <p>&nbsp;</p>
            <label className="checkbox is-small" htmlFor="forceMinimumLevel">
              <input
                type="checkbox"
                id="forceMinimumLevel"
                placeholder="3"
                checked={_forceMinimumLevel}
                onChange={(e) => { setForceMinimumLevel(e.target.checked); }}
              />
              &nbsp;Forçar nível mínimo
            </label>
          </div>
        </div>
        <div className="columns is-paddingless is-marginless">
          <div className="field column is-half">
            <label className="label is-small" htmlFor="mustIncludeNodes">Deve conter os nós
              <div className="control">
                <input
                  className="input is-small"
                  type="text"
                  id="mustIncludeNodes"
                  placeholder="AAA001,BBB002"
                  value={_mustIncludeNodes}
                  onChange={(e) => { setMustIncludeNodes(e.target.value); }}
                />
              </div>
            </label>
          </div>
          <div className="field column is-half">
            <label className="label is-small" htmlFor="mustNotIncludeSlots">Não deve conter os slots
              <div className="control">
                <input
                  className="input is-small"
                  type="text"
                  id="mustNotIncludeSlots"
                  placeholder="1,2,3"
                  value={_mustNotIncludeSlots}
                  onChange={(e) => { setMustNotIncludeSlots(e.target.value); }}
                />
              </div>
            </label>
          </div>
        </div>
        <div className="columns is-paddingless is-marginless">
          <div className="field column is-full">
            <button
              className="button is-primary is-small"
              type="button"
              onClick={() => {
                beginCalcNexts(graph, {
                  minWeight: _minWeight,
                  maxWeight: _maxWeight,
                  maxDistance: _maxDistance,
                  forceMinimumLevel: _forceMinimumLevel,
                  mustIncludeNodes: _mustIncludeNodes.length !== 0 ? _mustIncludeNodes.split(',') : null,
                  mustNotIncludeSlots: _mustNotIncludeSlots.length !== 0 ? _mustNotIncludeSlots.split(',').map((o) => parseInt(o, 10)) : null,
                }, (options) => setShowNextsOptions(Array(options.length).fill(false)));
              }}
            >
              Calcular
            </button>
          </div>
        </div>
      </div>
    );
  }

  return (
    <div className="modal is-active">
      <div className="modal-background" />
      <div className="modal-card">
        <header className="modal-card-head">
          <h1 className="title no-margin">Próximas</h1>
          <button
            type="button"
            className="delete"
            aria-label="close"
            onClick={hideNexts}
          />
        </header>
        <section className="modal-card-body">
          {content}
        </section>
        <footer className="modal-card-foot" />
      </div>
    </div>
  );
};

NextsOptions.propTypes = {
  nexts: nextsType.isRequired,
  graph: graphType.isRequired,
  showNexts: PropTypes.func.isRequired,
  hideNexts: PropTypes.func.isRequired,
  beginCalcNexts: PropTypes.func.isRequired,
  changeNodesStatus: PropTypes.func.isRequired,
};

export default NextsOptions;
