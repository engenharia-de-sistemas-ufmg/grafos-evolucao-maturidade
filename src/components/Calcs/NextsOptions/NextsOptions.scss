.inner-ul {
  padding-left: 12px;
}

.card {
  margin-bottom: 16px;
}

.card-header {
  cursor: pointer;
}

.card-header:hover {
  background-color: rgba(10, 10, 10, 0.2);
}

.card-header-subtitle {
  display: flex;
  flex-grow: 0;
  padding: 0.75rem;
}

.delete {
  -webkit-appearance: none;
  background-color: rgba(10, 10, 10, 0.2);
  border: none;
  border-radius: 290486px;
  cursor: pointer;
  pointer-events: auto;
  display: inline-block;
  flex-grow: 0;
  flex-shrink: 0;
  font-size: 0;
  height: 20px;
  max-height: 20px;
  max-width: 20px;
  min-height: 20px;
  min-width: 20px;
  outline: 0;
  position: relative;
  vertical-align: top;
  width: 20px;
  margin-left: auto;
}

.delete::before {
  height: 2px;
  width: 50%;
}

.delete::after {
  height: 50%;
  width: 2px;
}

.delete::before,
.delete::after {
  background-color: white;
  content: '';
  display: block;
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translateX(-50%) translateY(-50%) rotate(45deg);
  transform-origin: center center;
}
