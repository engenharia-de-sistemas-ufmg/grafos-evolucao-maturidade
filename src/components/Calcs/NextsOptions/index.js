import { connect } from 'react-redux';
import NextsOptions from './NextsOptions';
import { hideNexts } from '../../../actions/nexts';

const mapStateToProps = (state) => ({
  nexts: state.nexts,
});

const mapDispatchToProps = (dispatch) => ({
  hideNexts: () => { dispatch(hideNexts()); },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(NextsOptions);
