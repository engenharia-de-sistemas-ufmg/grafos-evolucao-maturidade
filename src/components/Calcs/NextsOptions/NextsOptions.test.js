import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import NextsOptions from './NextsOptions';
import basicGraph from '../../../examples/basicGraph';

describe('NextsOptions', () => {
  Enzyme.configure({ adapter: new Adapter() });

  test('should match snapshot', () => {
    const component = shallow(
      <NextsOptions
        nexts={{}}
        graph={basicGraph}
        showNexts={() => { }}
        hideNexts={() => { }}
        beginCalcNexts={() => { }}
        changeNodesStatus={() => { }}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
