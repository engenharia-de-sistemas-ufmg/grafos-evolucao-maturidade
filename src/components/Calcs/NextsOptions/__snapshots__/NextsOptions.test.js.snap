// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextsOptions should match snapshot 1`] = `
<div
  className="modal is-active"
>
  <div
    className="modal-background"
  />
  <div
    className="modal-card"
  >
    <header
      className="modal-card-head"
    >
      <h1
        className="title no-margin"
      >
        Próximas
      </h1>
      <button
        aria-label="close"
        className="delete"
        onClick={[Function]}
        type="button"
      />
    </header>
    <section
      className="modal-card-body"
    >
      <div
        className="content"
      >
        <div
          className="columns is-paddingless is-marginless"
        >
          <div
            className="field column is-one-quarter"
          >
            <label
              className="label is-small"
              htmlFor="minWeight"
            >
              Peso mínimo
              <div
                className="control"
              >
                <input
                  className="input is-small"
                  id="minWeight"
                  onChange={[Function]}
                  placeholder="195"
                  type="text"
                  value={195}
                />
              </div>
            </label>
          </div>
          <div
            className="field column is-one-quarter"
          >
            <label
              className="label is-small"
              htmlFor="maxWeight"
            >
              Peso máximo
              <div
                className="control"
              >
                <input
                  className="input is-small"
                  id="maxWeight"
                  onChange={[Function]}
                  placeholder="375"
                  type="text"
                  value={375}
                />
              </div>
            </label>
          </div>
          <div
            className="field column is-one-quarter"
          >
            <label
              className="label is-small"
              htmlFor="maxDistance"
            >
              Distância Máxima
              <div
                className="control"
              >
                <input
                  className="input is-small"
                  id="maxDistance"
                  onChange={[Function]}
                  placeholder="3"
                  type="text"
                  value={3}
                />
              </div>
            </label>
          </div>
          <div
            className="field column is-one-quarter"
          >
            <p>
               
            </p>
            <label
              className="checkbox is-small"
              htmlFor="forceMinimumLevel"
            >
              <input
                checked={true}
                id="forceMinimumLevel"
                onChange={[Function]}
                placeholder="3"
                type="checkbox"
              />
               Forçar nível mínimo
            </label>
          </div>
        </div>
        <div
          className="columns is-paddingless is-marginless"
        >
          <div
            className="field column is-half"
          >
            <label
              className="label is-small"
              htmlFor="mustIncludeNodes"
            >
              Deve conter os nós
              <div
                className="control"
              >
                <input
                  className="input is-small"
                  id="mustIncludeNodes"
                  onChange={[Function]}
                  placeholder="AAA001,BBB002"
                  type="text"
                  value=""
                />
              </div>
            </label>
          </div>
          <div
            className="field column is-half"
          >
            <label
              className="label is-small"
              htmlFor="mustNotIncludeSlots"
            >
              Não deve conter os slots
              <div
                className="control"
              >
                <input
                  className="input is-small"
                  id="mustNotIncludeSlots"
                  onChange={[Function]}
                  placeholder="1,2,3"
                  type="text"
                  value=""
                />
              </div>
            </label>
          </div>
        </div>
        <div
          className="columns is-paddingless is-marginless"
        >
          <div
            className="field column is-full"
          >
            <button
              className="button is-primary is-small"
              onClick={[Function]}
              type="button"
            >
              Calcular
            </button>
          </div>
        </div>
      </div>
    </section>
    <footer
      className="modal-card-foot"
    />
  </div>
</div>
`;
